keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,344 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The main goal of this enhancement is to authoritatively identify Windows
+pods during the API Server admission time. 
+
+
+## Motivation
+
+Identifying Windows pods during the API Server admission time is crucial to apply
+appropriate security constraints to the pod. Without proper identification, some admission
+plugins may apply unnecessary security constraints to the Windows pods or in the
+worst case, don't apply security constraints at all.
+
+
+### Goals
+
+Admission plugins and/or Admission Webhooks identifying if the incoming pod
+is Windows or not.
+
+
+### Non-Goals
+
+
+- Validating if the Windows pods have appropriate security constraints.
+- Interaction with container runtimes.
+
+## Proposal
+
+
+As mentioned [earlier](#Motivation) identifying the Windows pods during the scheduling phase can be done
+in the following ways:
+- Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 
+  Windows node specific labels.
+- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec
+
+The problem with using nodeSelector and tolerations is that any unprivileged user/entity can apply the 
+nodeSelector and/or tolerations in the pod spec making it unsecure whereas RuntimeClasses are [recommended](https://kubernetes.io/docs/concepts/containers/runtime-class/#setup) to be
+created by cluster administrator making them authoritative enough to be used.
+
+### User Stories (Optional)
+
+
+#### Story 1
+
+As a Kubernetes cluster administrator, I want appropriate security contexts to be 
+applied to my Windows Pods along with Linux pods
+
+#### Story 2
+
+As a Kubernetes cluster administrator, I want to use my own admission webhook for
+Windows pods.
+
+### Notes/Constraints/Caveats (Optional)
+
+While the `scheduling` field in the RuntimeClass is created to handle scheduling constraints, we have a reserved label
+[`kubernetes.io/os`](https://kubernetes.io/docs/reference/labels-annotations-taints/#kubernetes-io-os) to identify
+Windows nodes from their Linux counterparts. If the RuntimClass's scheduling field has a nodeSelector  `kubernetes.io/os: 'windows'`,
+we can authoritatively say that the pod is indeed a Windows one during the api-server admission time. Querying the RuntimeClass
+during admission time may be expensive operation as the admission plugin is not stateless anymore. 
```

Comment:
How are you going to mitigate or quantify this expense?

Reply:
It will be amortized if we go with runtimeclass admission plugin as it is making an apicall to get , if not it'll be O(1) per pod(the admission works per pod) as pod spec can have maximum of 1 runtimeclass * No of pod creations in the cluster.(Need to check if we allow updating runtimeclasses which I think we don't do because pod spec is immutable).

Reply:
I suggest calling it out here

Reply:
Scheduling constraints on the RuntimeClass are merged into the PodSpec during admission (mutating), so I don't think an admission controller would actually need to query the runtimeclass.

Reply:
That's why I said, if we go with runtimeclass admission plugin we won't need querying as it is already making an api call. However, we'd also want the same mechanism to be used in pod security admission plugin to identify Windows pods and for that we will need to query runtimeclasses unless you're suggesting something different.

Reply:
What I mean is, validating admission happens after mutating admission, so as long as you're using the runtimeclass admission plugin (which is expected, if using RuntimeClasses) then the information will already be merged into the pod by the time the PodSecurity admission plugin runs.

Reply:
To be clear, if we go with this approach, the runtimeclasses admission plugin is not only going to mutate, it is also going to do a validation of the pod.

Sample [PR](https://github.com/kubernetes/kubernetes/pull/103784/files) which does this.

Reply:
The responsibilities and actors in this doc are hard to follow... I *think* this is proposing the following:

Determining windows pods in validating admission plugins:

1. make pod `spec.nodeSelector` `kubernetes.io/os: windows` definitive for determining a pod is a Windows pod
2. make kubelet refuse to run pods that have a `kubernetes.io/os` spec.nodeSelector that does not match the kubelet's runtime.GOOS

Mechanisms for setting spec.nodeSelector:

1. user can set spec.nodeSelector in workload template
2. user can set `spec.runtimeClassName` in workload template, admin can set `scheduling:{nodeSelector:{"kubernetes.io/os":"windows"}}` in runtimeclass, runtimeclass mutating admission copies into pod `spec.nodeSelector` during mutating admission, ensuring validating admission plugins have the pod spec.nodeSelector data available to inspect

Distinguishing the "what identifies a windows pod?" details from the "how does that data get set in the pod API?" details will make this proposal easier to reason about.

Reply:
One implication of having the runtimeclass admission plugin apply the `kubernetes.io/os: windows` nodeSelector at pod creation time is that it leaves the workload resource pod templates without a `kubernetes.io/os: windows` nodeSelector. That means that admission plugins that inspect the pod templates in those workload resources will not know what OS the resulting pods will target.

As a concrete example, if you set up PodSecurityAdmission to warn users if their deployments don't fit inside the `restricted` Pod Security Standards level, and their deployment targeted a windows runtimeClassName but didn't set a `kubernetes.io/os: windows` nodeSelector and didn't set `allowPrivilegeEscalation: false`, the user would get warned that the pod template in the deployment didn't fit within the `restricted` policy level.

Reply:
> > One implication of having the runtimeclass admission plugin apply the kubernetes.io/os: windows nodeSelector at pod creation time is that it leaves the workload resource pod templates without a kubernetes.io/os: windows nodeSelector.
> 
> That is good point. Isn't that an existing problem with runtimeclass admission plugin

workload objects targeting a windows runtimeClassName which lack a windows node selector aren't an issue because nothing in in-tree admission uses node selectors as an input for warning/validating the pod template.

> * Have runtimeclass admission plugin also update the workload template's nodeSelector and tolerations instead of just updating the pod object so that the validation can be smooth in PodSecurityAdmission plugin

I don't think this is a good idea. Pod object selectors are immutable once set... it is significantly harder to correctly merge user- and admission-specified selectors on objects which can be updated.

> * Have the same logic of querying runtimeclasses for the workload templates/pods in PodSecurityAdmission plugin and instead warn for both workload templates/pods.

I also don't think PodSecurity admission depending on current state of RuntimeClass objects is a good idea... that admission plugin works hard to keep the evaluation of a given policy level against a given pod not depend on the state of anything else in the cluster.

Reply:
> workload objects targeting a windows runtimeClassName which lack a windows node selector aren't an issue because nothing in in-tree admission uses node selectors as an input for warning/validating the pod template.

Correct. That's why I said compounded by validating presence `kubernetes.io/os` label. 

> I also don't think PodSecurity admission depending on current state of RuntimeClass objects is a good idea... that admission plugin works hard to keep the evaluation of a given policy level against a given pod not depend on the state of anything else in the cluster.


I agree with the first part that it is significantly hard to update Pod object selectors in a mutating admission plugin however if PodSecurity admission plugin is anyways allowing certain runtimeclasses to excluded statically via config, how difficult would it to be to do the same dynamically. I know that we prefer admission plugins to be stateless however since PodSecurity admission plugin is not just validating pods and it is validating workload templates and we cannot update the Pod templates in workload spec, I believe there are no alternatives to having the logic in PodSecurity admission plugin.

Reply:
Had an offline discussion with @liggitt. The main points are:
- Have the nodeSelector explicitly set by end-user in the workload templates so that PodSecurityAdmission plugin can ignore workloads that have windows nodeSelector set.
- Let kubelet check for the nodeSelector and see if it matches node.Labels[` kubernetes.io/os `] instead of checking `runtime.GOOS` directly.

I am fine with the approach suggested. The only downside I can think of above is since nodeSelector can be added by anyone(not governed by RBAC or quota), can it be trusted during API admission time?

Reply:
The selector is as trustworthy as the labels on the node are today. The two actors that pay attention to the selector are the scheduler and the kubelet. If the change to the kubelet to additionally verify the selector against labels forced to match the actual OS of the node, that closes the only gap I can see (if the Node API object label did not match the node OS)

Reply:
Won't we also make PodSecurityAdmission plugin trust the nodeSelector label during API admission time to differentiate Windows workloads from Linux?

Reply:
once we know the node will refuse to run a pod with an OS selector that does not match the actual OS, that seems ok

Reply:
there is a symmetry issue we should call out:

>make pod spec.nodeSelector kubernetes.io/os: windows definitive for determining a pod is a Windows pod

it is also authoritative to determine if a pod is on a linux host.

for a broader ecosystem of webhooks/validation/injectors, is it easier to parse pod.spec.nodeSelector or easier to just have a dedicated field that says this pod requires this OS.